prompting/storage/storage.go

```diff
@@ -5,20 +5,42 @@ import (
 	"errors"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/prompting/notifier"
 )
 
 var ErrNoSavedDecision = errors.New("no saved prompt decision")
-
-type userDB struct {
-	PerLabelDB map[string]*labelDB `json:"per-label-db"`
-}
+var ErrMultipleDecisions = errors.New("multiple prompt decisions for the same path")
+
+const (
```

Comment:
A custom type could be used here to make things a bit more type safe (and also arguably slightly easier to read i nthe code). Something like (all names strawmans, I really am short on time to review this properly sadly)
```go
diff --git a/prompting/storage/storage.go b/prompting/storage/storage.go
index 3c16df1eee..5fd51ea0ef 100644
--- a/prompting/storage/storage.go
+++ b/prompting/storage/storage.go
@@ -15,12 +15,18 @@ import (
 var ErrNoSavedDecision = errors.New("no saved prompt decision")
 var ErrMultipleDecisions = errors.New("multiple prompt decisions for the same path")
 
+type jsonAllowTypes string
+
 const (
        // must match the json annotations for entries in labelDB
-       jsonAllow            = "allow"
-       jsonAllowWithDir     = "allow-with-dir"
-       jsonAllowWithSubdirs = "allow-with-subdir"
+       jsonAllowUnset       jsonAllowTypes = ""
+       jsonAllow            jsonAllowTypes = "allow"
+       jsonAllowWithDir     jsonAllowTypes = "allow-with-dir"
+       jsonAllowWithSubdirs jsonAllowTypes = "allow-with-subdir"
        // must match the specification for extra information map returned by the prompt
+)
+
+const (
        extrasAlwaysPrompt     = "always-prompt"
        extrasAllowWithDir     = "allow-directory"
        extrasAllowWithSubdirs = "allow-subdirectories"
@@ -56,7 +62,7 @@ func New() *PromptsDB {
        return pd
 }
 
-func findPathInLabelDB(db *labelDB, path string) (bool, error, string, string) {
+func findPathInLabelDB(db *labelDB, path string) (bool, error, jsonAllowTypes, string) {
        // Returns:
        // bool: allow
        // error: (nil | ErrMultipleDecisions | ErrNoSavedDecision)
@@ -64,7 +70,7 @@ func findPathInLabelDB(db *labelDB, path string) (bool, error, string, string) {
        // string: matching path current in the db
        path = filepath.Clean(path)
        storedAllow := true
-       which := ""
+       which := jsonAllowUnset
        var err error
        // Check if original path has exact match in db.Allow
        if allow, exists := db.Allow[path]; exists {
@@ -161,7 +167,7 @@ func (pd *PromptsDB) load() error {
        return json.NewDecoder(f).Decode(&pd.PerUser)
 }
 
-func whichMap(allow bool, extras map[string]string) string {
+func whichMap(allow bool, extras map[string]string) jsonAllowTypes {
        if (allow && extras[extrasAllowWithSubdirs] == "yes") || (!allow && extras[extrasDenyWithSubdirs] == "yes") {
                return jsonAllowWithSubdirs
        }
```

Reply:
Thanks for the excellent suggestion! This was what I was attempting to do wasn't sure how idiomatically in go. I'll make this change in a future PR.